src/Bridges/Constraint/bridges/sos1_to_milp.jl

```diff
@@ -0,0 +1,254 @@
+
+"""
+    SOS1ToMILPBridge{T,F} <: Bridges.Constraint.AbstractBridge
+
+`SOS1ToMILPBridge` implements the following reformulation:
+
+  * ``x \\in \\textsf{SOS1}(d)`` into a mixed-integer linear program.
+
+## Source node
+
+`SOS1ToMILPBridge` supports:
+
+  * `F` in [`MOI.SOS1`](@ref)
+
+where `F` is [`MOI.VectorOfVariables`](@ref) or
+[`MOI.VectorAffineFunction{T}`](@ref).
+
+## Target nodes
+
+`SOS1ToMILPBridge` creates:
+
+  * [`MOI.VariableIndex`](@ref) in [`MOI.ZeroOne`](@ref)
+  * [`MOI.ScalarAffineFunction{T}`](@ref) in [`MOI.EqualTo{T}`](@ref)
+  * [`MOI.ScalarAffineFunction{T}`](@ref) in [`MOI.LessThan{T}`](@ref)
+"""
+mutable struct SOS1ToMILPBridge{
+    T,
+    F<:Union{MOI.VectorOfVariables,MOI.VectorAffineFunction{T}},
+} <: AbstractBridge
+    f::F
+    s::MOI.SOS1{T}
+    variables::Vector{MOI.VariableIndex}
+    # ∑_i z_i == 1
+    equal_to::MOI.ConstraintIndex{MOI.ScalarAffineFunction{T},MOI.EqualTo{T}}
+    # x_i - u_i z_i <= 0 ∀i
+    # l_i z_i - x_i <= 0 ∀i
+    less_than::Vector{
+        MOI.ConstraintIndex{MOI.ScalarAffineFunction{T},MOI.LessThan{T}},
+    }
+    bounds::Vector{NTuple{2,T}}
+    function SOS1ToMILPBridge{T}(
+        f::Union{MOI.VectorOfVariables,MOI.VectorAffineFunction{T}},
+        s::MOI.SOS1{T},
+    ) where {T}
+        return new{T,typeof(f)}(
+            f,
+            s,
+            MOI.VariableIndex[],
+            MOI.ConstraintIndex{MOI.ScalarAffineFunction{T},MOI.EqualTo{T}}(0),
+            MOI.ConstraintIndex{MOI.ScalarAffineFunction{T},MOI.LessThan{T}}[],
+            NTuple{2,T}[],
+        )
+    end
+end
+
+const SOS1ToMILP{T,OT<:MOI.ModelLike} =
+    SingleBridgeOptimizer{SOS1ToMILPBridge{T},OT}
+
+function bridge_constraint(
+    ::Type{SOS1ToMILPBridge{T,F}},
+    model::MOI.ModelLike,
+    f::F,
+    s::MOI.SOS1,
+) where {T,F<:Union{MOI.VectorOfVariables,MOI.VectorAffineFunction{T}}}
+    # !!! info
+    #     Postpone rest of creation until final_touch.
+    return SOS1ToMILPBridge{T}(f, s)
+end
+
+function MOI.supports_constraint(
+    ::Type{<:SOS1ToMILPBridge{T}},
+    ::Type{<:Union{MOI.VectorOfVariables,MOI.VectorAffineFunction{T}}},
+    ::Type{MOI.SOS1{T}},
+) where {T}
+    return true
+end
+
+function MOI.Bridges.added_constrained_variable_types(
+    ::Type{<:SOS1ToMILPBridge},
+)
+    return Tuple{Type}[(MOI.ZeroOne,)]
+end
+
+function MOI.Bridges.added_constraint_types(
+    ::Type{<:SOS1ToMILPBridge{T}},
+) where {T}
+    return Tuple{Type,Type}[
+        (MOI.ScalarAffineFunction{T}, MOI.EqualTo{T}),
+        (MOI.ScalarAffineFunction{T}, MOI.LessThan{T}),
+    ]
+end
+
+function concrete_bridge_type(
+    ::Type{<:SOS1ToMILPBridge{T}},
+    ::Type{F},
+    ::Type{MOI.SOS1{T}},
+) where {T,F<:Union{MOI.VectorOfVariables,MOI.VectorAffineFunction{T}}}
+    return SOS1ToMILPBridge{T,F}
+end
+
+function MOI.get(
+    ::MOI.ModelLike,
+    ::MOI.ConstraintFunction,
+    bridge::SOS1ToMILPBridge,
+)
+    return bridge.f
```

Comment:
You might need to copy the function here for the same reason `MOI.Utilities.Model` does a copy. IIRC, the user is allowed to modify the function provided by `add_constraint`. If you don't make a copy, it will modify the function stored in the bridge.